// payload/dist/graphql/schema/buildPoliciesType.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_type_json_1 = require("graphql-type-json");
const graphql_1 = require("graphql");
const formatName_1 = __importDefault(require("../utilities/formatName"));
const buildFields = (label, fieldsToBuild) => fieldsToBuild.reduce((builtFields, field) => {
    if (!field.hidden) {
        if (field.name) {
            const fieldName = (0, formatName_1.default)(field.name);
            const objectTypeFields = ['create', 'read', 'update', 'delete'].reduce((operations, operation) => {
                const capitalizedOperation = operation.charAt(0).toUpperCase() + operation.slice(1);
                return {
                    ...operations,
                    [operation]: {
                        type: new graphql_1.GraphQLObjectType({
                            name: `${label}_${fieldName}_${capitalizedOperation}`,
                            fields: {
                                permission: {
                                    type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean),
                                },
                            },
                        }),
                    },
                };
            }, {});
            if (field.fields) {
                objectTypeFields.fields = {
                    type: new graphql_1.GraphQLObjectType({
                        name: `${label}_${fieldName}_Fields`,
                        fields: buildFields(`${label}_${fieldName}`, field.fields),
                    }),
                };
            }
            return {
                ...builtFields,
                [field.name]: {
                    type: new graphql_1.GraphQLObjectType({
                        name: `${label}_${fieldName}`,
                        fields: objectTypeFields,
                    }),
                },
            };
        }
        if (!field.name && field.fields) {
            const subFields = buildFields(label, field.fields);
            return {
                ...builtFields,
                ...subFields,
            };
        }
        if (field.type === 'tabs') {
            return field.tabs.reduce((fieldsWithTabFields, tab) => {
                return {
                    ...fieldsWithTabFields,
                    ...buildFields(label, tab.fields),
                };
            }, { ...builtFields });
        }
    }
    return builtFields;
}, {});
const buildEntity = (label, entityFields, operations) => {
    const formattedLabel = (0, formatName_1.default)(label);
    const fields = {
        fields: {
            type: new graphql_1.GraphQLObjectType({
                name: (0, formatName_1.default)(`${formattedLabel}Fields`),
                fields: buildFields(`${formattedLabel}Fields`, entityFields),
            }),
        },
    };
    operations.forEach((operation) => {
        const capitalizedOperation = operation.charAt(0).toUpperCase() + operation.slice(1);
        fields[operation] = {
            type: new graphql_1.GraphQLObjectType({
                name: `${formattedLabel}${capitalizedOperation}Access`,
                fields: {
                    permission: { type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean) },
                    where: { type: graphql_type_json_1.GraphQLJSONObject },
                },
            }),
        };
    });
    return fields;
};
function buildPoliciesType(payload) {
    const fields = {
        canAccessAdmin: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean),
        },
    };
    Object.values(payload.config.collections).forEach((collection) => {
        const collectionOperations = ['create', 'read', 'update', 'delete'];
        if (collection.auth && (typeof collection.auth.maxLoginAttempts !== 'undefined' && collection.auth.maxLoginAttempts !== 0)) {
            collectionOperations.push('unlock');
        }
        if (collection.versions) {
            collectionOperations.push('readVersions');
        }
        fields[(0, formatName_1.default)(collection.slug)] = {
            type: new graphql_1.GraphQLObjectType({
                name: (0, formatName_1.default)(`${collection.labels.singular}Access`),
                fields: buildEntity(collection.labels.singular, collection.fields, collectionOperations),
            }),
        };
    });
    Object.values(payload.config.globals).forEach((global) => {
        const globalOperations = ['read', 'update'];
        if (global.versions) {
            globalOperations.push('readVersions');
        }
        fields[(0, formatName_1.default)(global.slug)] = {
            type: new graphql_1.GraphQLObjectType({
                name: (0, formatName_1.default)(`${global.label}Access`),
                fields: buildEntity(global.label, global.fields, globalOperations),
            }),
        };
    });
    return new graphql_1.GraphQLObjectType({
        name: 'Access',
        fields,
    });
}
exports.default = buildPoliciesType;
